File: Crio/crio_programming_interview_problems-master/MinCostOfRoad/MinCostOfRoad.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;


// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
bool comp(const vector<int> &a,const vector<int> &b){
    return a[2]<b[2];
}
int getParent(int x, vector<int> &parent)
{
    if(parent[x]==x)return x;

    return parent[x]=getParent(parent[x],parent);
}

void setUnion(int a, int b, vector<int> &parent, vector<int> &size)
{
    a = getParent(a, parent);
    b = getParent(b, parent);

    if (size[a] > size[b])
    {
        swap(a, b);
    }

    parent[b] = a;
    size[a] += size[b];
    return;
}
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

int minCostOfRoad(int n, vector<vector<int> > &edges)
{
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    sort(edges.begin(), edges.end(),comp);
    vector<int> size(n, 0);

    ll cost = 0;

    vector<int> parent(n+1);
    for (int i = 0; i <= n; i++)
    {
        parent[i] = i;
    }

    for (auto edge : edges)
    {
        int p1 = getParent(edge[0], parent);
        int p2 = getParent(edge[1], parent);

        if (p1 == p2)
        {
            continue;
        }
        cost += edge[2];
        setUnion(p1, p2, parent, size);
    }

    return cost;
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}

int main()
{
    int n,m;
    cin >> n>>m;

    vector<vector<int> > edges;

    for (int i = 0; i < m; i++)
    {   int u,v,w;cin>>u>>v>>w;
        vector<int> b(3,0);
        b[0]=u;
        b[1]=v;
        b[2]=w;
        edges.push_back(b);
    }

    int ans = minCostOfRoad(n, edges);

    cout << ans << endl;
}

File: Crio/crio_programming_interview_problems-master/MinCostOfRoad/MinCostOfRoad.py
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
parent= []
def init(n):
    for i in range(n+1):
        parent.append(i)
def findParent(x):
    while x!= parent[x]:
        x=parent[parent[x]]
    return x
def union(x,y):
    p=findParent(x)
    q=findParent(y)
    if p!= q:
        parent[p]=q
 # CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
def minCostOfRoad(n,edges):
    # CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    edges=sorted(edges,key=lambda item: item[2])
    e=len(edges)
    cnt =0
    cost=0
    init(n)
    for i in range(e):
        x=findParent(edges[i][0])
        y=findParent(edges[i][1])
        if x != y:
            cnt+=1
            cost += edges[i][2]
            union(edges[i][0],edges[i][1])

    return cost
    # CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def main():
    n,e=map(int,input().split())
    edges=[]
    for i in range(e):
        a,b,c= map(int,input().split())
        edges.append([a,b,c])
    ans = minCostOfRoad(n,edges)
    print(ans)
if __name__ == '__main__':
    main()
